### PR TITLE
Feature: first design console

### PR DESCRIPTION
**Pull request by Olavhaasie**
Closes PM-29

## Description
This PR adds a design for the minute taking Qt plugin. It provides a UI that can be used to enter messages. The only functionality that has been implemented is entering messages, it currently does not listen to any topics.

image

This can replace the `march_rqt_training_log` functionality. This plugin should filter messages and topics and publish messages relevant to the training process. See PM-15 for more information.

## Changes
* Added `march_rqt_console` package
* Added basic `console.ui`

### CODE REVIEW

**Review by RutgerVanBeek — COMMENTED** [review]
How would you add your message to the console? Shouldn't there be a button for this. 
What would you load and save with the buttons ?
How would you delete a message you added?

**Comment by Olavhaasie** [comment]
> How would you add your message to the console? Shouldn't there be a button for this.

In the input field at the bottom you can type your minutes and press enter to submit them. I wanted to keep the GUI clean, so I left out a button.

> What would you load and save with the buttons ?

You could save all the current messages to a file in a nice format or in csv.

> How would you delete a message you added?

This does not work yet, but you can select the messages from the table and delete them with your delete key or we could implement a right click menu with more options.

**Review by RutgerVanBeek — DISMISSED** [review]

File: march_rqt_console/src/march_rqt_console/message.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+from python_qt_binding.QtCore import QDateTime, QObject, QTime
+
+
+class Message(QObject):
```
Comment (Roelemans) [inline march_rqt_console/src/march_rqt_console/message.py]: Shouldn't you better call this LogEntry or something to avoid confusion with ros messages?

**Review by Roelemans — COMMENTED** [review]
Why is this called console? I would interpret that as a menu from which you could launch other rqt plugins and the like. Maybe call it note_taker/notes/minutes/minute_taker? Why do you add a "level" column? I think only "error/fatal" would be a relevant level, but you might as well just include that in the message (which you probably will anyway). Perhaps you could use the column to add labels like "error/fatal" and (emergency button) "pressed", "todo" etc. 

How exactly would the camera take work? Does it send a message when you change the take?

**Comment by Olavhaasie** [comment]
> Why is this called console? I would interpret that as a menu from which you could launch other rqt plugins and the like. Maybe call it note_taker/notes/minutes/minute_taker?

I copied the name from `rqt_console` and forgot to give it a descriptive name, so good point, I will call it... `march_note_taker` :confetti_ball: 

> Why do you add a "level" column? I think only "error/fatal" would be a relevant level, but you might  as well just include that in the message (which you probably will anyway). Perhaps you could use the column to add labels like "error/fatal" and (emergency button) "pressed", "todo" etc.

I agree, we can give the logs a color like red for error/fatal and just gray for info. And use the column for tags.

> How exactly would the camera take work? Does it send a message when you change the take?

Precisely my idea :innocent: Do you have any other ideas?

**Comment by Roelemans** [comment]
> Precisely my idea :innocent: Do you have any other ideas?

I think it is a bit strange to have a take "scroller" like you have here (if you scroll it would send multiple messages). Also, sometimes you jump from take 3 to take 7 due to failed takes. Maybe either
1) add an "add camera take" button (or shortcut)? 
2) when you change the camera take, it adds a message to the text bar for logging; which is replaced when you change it again (so you would still have to press enter to send the message). Would have to think about what to do when there is already an unfinished message present in the text bar
3) Add a message to the log which is replaced when the take is changed again within a certain time slot/or so long as no custom message is send

**Comment by Olavhaasie** [comment]
> add an "add camera take" button (or shortcut)?

Then I would go with your first idea with button as well as shortcut. Since that is easy to implement and use.

File: march_rqt_note_taker/src/march_rqt_note_taker/notes_plugin.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+        self._widget.table_view.setModel(self._model)
+        self._widget.table_view.verticalScrollBar().rangeChanged.connect(self.change_scroll)
+
+        self._widget.input_field.returnPressed.connect(self.insert_entry)
+
+        self._widget.take_button.clicked.connect(self.start_take)
+
+    def init_ui(self, context):
+        ui_file = os.path.join(rospkg.RosPack().get_path('march_rqt_note_taker'), 'resource', 'plugin.ui')
```
Comment (Roelemans) [inline march_rqt_note_taker/src/march_rqt_note_taker/notes_plugin.py]:
```suggestion
        ui_file = os.path.join(rospkg.RosPack().get_path('march_rqt_note_taker'), 'resource', 'note_taker.ui')
```

File: march_rqt_note_taker/resource/plugin.ui [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+  <property name="geometry">
+   <rect>
+    <x>0</x>
+    <y>0</y>
+    <width>838</width>
+    <height>675</height>
+   </rect>
+  </property>
+  <property name="windowTitle">
+   <string>Console</string>
```
Comment (Roelemans) [inline march_rqt_note_taker/resource/plugin.ui]:
```suggestion
   <string>Note taker</string>
```

File: march_rqt_note_taker/plugin.xml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<library path="src">
+    <class name="NotesPlugin" type="march_rqt_note_taker.notes_plugin.NotesPlugin" base_class_type="rqt_gui_py::Plugin">
+        <description>
+            An RQT plugin for monitoring log messages and writing logs.
+        </description>
+        <qtgui>
+            <label>March Minutes</label>
```
Comment (Roelemans) [inline march_rqt_note_taker/plugin.xml]:
```suggestion
            <label>March note taker</label>
```

File: march_rqt_note_taker/plugin.xml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+            <icon>resource/img/march-walking.png</icon>
+            <statustip>March Minutes</statustip>
```
Comment (Roelemans) [inline march_rqt_note_taker/plugin.xml]:
```suggestion
            <statustip>March note taker</statustip>
```

**Review by Roelemans — CHANGES_REQUESTED** [review]

File: march_rqt_note_taker/resource/note_taker.ui [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  <property name="geometry">
+   <rect>
+    <x>0</x>
+    <y>0</y>
+    <width>838</width>
+    <height>675</height>
+   </rect>
+  </property>
+  <property name="windowTitle">
+   <string>Console</string>
```
Comment (Roelemans) [inline march_rqt_note_taker/resource/note_taker.ui]:
```suggestion
   <string>Note taker</string>
```

**Review by Roelemans — CHANGES_REQUESTED** [review]

**Review by Roelemans — APPROVED** [review]
Took you long enough

**Pull request merged**